examples/classic/distributed/Paxos/Paxos.qnt

```diff
@@ -34,15 +34,41 @@ module Paxos {
   var maxVal:   str -> int  // ballot number cast by a; it equals <<-1, None>> if
                             // a has not cast any vote.
 
-  // A union type for records representing messages.
+  // Particular message types we need to refer to later:
+  type Msg1bT = { acceptor: str, bal: int, mbal: int, mval: int }
+  type Msg2aT = { bal: int, value: int }
+
+  // A union type for records representing different types of messages.
   type MSG =
-      | { tag: "1a", bal: int }
-      | { tag: "1b", acceptor: str, bal: int, mbal: int, mval: int }
-      | { tag: "2a", bal: int, value: int }
-      | { tag: "2b", acc: str, bal: int, value: int }
+      | Msg1a({ bal: int })
+      | Msg1b(Msg1bT)
+      | Msg2a(Msg2aT)
+      | Msg2b({ acc: str, bal: int, value: int })
 
   var msgs: Set[MSG] // The set of all messages that have been sent.
 
+  // Access the balance of a message, irrespictive of the message variant
+  def balance(m: MSG): int = match m {
+    | Msg1a(mm) => mm.bal
+    | Msg1b(mm) => mm.bal
+    | Msg2a(mm) => mm.bal
+    | Msg2b(mm) => mm.bal
+  }
+
+  // Filter a set of messager into just 1b messages
+  def filterMsg1b(ms: Set[MSG]): Set[Msg1bT] =
+    ms.fold(Set(), (acc, m) => match m {
+      | Msg1b(m1b) => acc.union(Set(m1b))
+      | _          => acc
+    })
+
+  // Filter a set of messager into just 2a messages
+  def filterMsg2a(ms: Set[MSG]): Set[Msg2aT] =
+    ms.fold(Set(), (acc, m) => match m {
+      | Msg2a(m2a) => acc.union(Set(m2a))
+      | _          => acc
+    })
```

Comment:
`fold` seems to be an overkill for me. Why would not we just use `filter`?

Reply:
We are mapping values out the type `Set[MSG]` into `Set[Msg2aT]`. There is no way of doing this with a filter, which can reduce the elements of the set, but cannot (given our current type system) reduce the type of the elements in the set. If we had instead exposed the extensibility of the extensible variants we are based on -- i.e, by allowing anonymous sum types -- we could give filter a typing that could refine the type of the collection it works on, in this case to just holding `Msg2a` variants. Without mapping out of `Set[MSG]`, whatever consumed the result of the filter would still need to account for all possible message types.